include/seqan3/argument_parser/validators.hpp

```diff
@@ -297,68 +290,347 @@ class file_ext_validator
-     * \param  v          The input range to iterate over and check every element.
-     * \throws parser_invalid_argument
+    //!\brief Returns a message that can be appended to the (positional) options help page info.
+    std::string get_help_page_message() const
+    {
+        return detail::to_string("An existing path for the input directory.");
+    }
+};
+
+/*!\brief A validator that checks if a given path is a valid output directory.
+ * \ingroup argument_parser
+ * \implements seqan3::validator_concept
+ *
+ * \details
+ *
+ * The struct acts as a functor that throws a seqan3::parser_invalid_argument exception whenever a given path
+ * (std::filesystem::path) is not writable. This can happen if either the parent path does not exists, or the
+ * path doesn't have the proper write permissions.
+ *
+ * \snippet test/snippet/argument_parser/validators_output_directory.cpp usage
+ *
+ * \note The validator works on every type that can be implicitly converted to std::filesystem::path.
+ */
+class output_directory_validator : public file_validator_base
+{
+public:
+    //!\brief Type of values that are tested by validator.
+    using file_validator_base::value_type;
+
+    /*!\name Constructors, destructor and assignment
+     * \{
      */
-    template <std::ranges::ForwardRange range_type>
-    //!\cond
-        requires std::ConvertibleTo<value_type_t<range_type>, std::filesystem::path const &>
-    //!\endcond
-    void operator()(range_type const & v) const
+    output_directory_validator() = default;                                               //!< Defaulted.
+    output_directory_validator(output_directory_validator const &) = default;             //!< Defaulted.
+    output_directory_validator(output_directory_validator &&) = default;                  //!< Defaulted.
+    output_directory_validator & operator=(output_directory_validator const &) = default; //!< Defaulted.
+    output_directory_validator & operator=(output_directory_validator &&) = default;      //!< Defaulted.
+    ~output_directory_validator() = default;                                              //!< Defaulted.
+
+    //!\brief Inherit base constructor.
+    using file_validator_base::file_validator_base;
+    //!\}
+
+    // Import the base::operator().
+    using file_validator_base::operator();
+
+    /*!\brief Tests whether path is writable.
+     * \param dir The input value to check.
+     * \throws parser_invalid_argument if the validation process failed or std::filesystem::filesystem_error if some
+     *         unexpected errors during the filesystem operations occur.
+     */
+    void operator()(std::filesystem::path const & dir) const override
     {
-         std::for_each(v.begin(), v.end(), [&] (auto cmp) { (*this)(cmp); });
+        std::function<void(std::filesystem::path const *)> deleter;
+        if (std::filesystem::exists(dir))
+            deleter = [] (std::filesystem::path const *) { /* no-op */ };
```

Comment:
isn't this a bit over the top? 
Couldn't you just do
```
bool existed = std::filesystem::exists(dir);

//... create dir and dummy file

if (!existed)
    std::filesystem::remove(*p, ec);
```
?

Reply:
In fact it is not over the top. All filesystem operations can possibly throw and this is a resource that needs to be cleaned up in case something went wrong. Accordingly, I wrap it in a RAII object to avoid unexpected results. Also it would be necessary to put the exists check in at least two different places.

Reply:
But you are using the filesystem overloads that are noexcept :thinking: ?

Reply:
not all. Only if the validator should capture this as an error. At least path_is_writable, ... oops I meant filename_is_writable 😈 , can throw std::filesystem_error. But even if not, to protect a resource you have to wrap it into RAII according to the CPP core guidelines.

Reply:
hm okay, I think making filename_is_writable noexcept would be good anyway. 

In your current solution, no error is thrown if the directory could not be deleted right? should this be the desired behaviour? Because then the directory might have been created but not removed... which might be weird behaviour for a user if the program fails afterward because the directory already exists (an the developer assumed it is not)?

Reply:
Well, something is really really broken if I can create the directory but not delete it. So this case will very likely never happen? Like when another process deletes the directory in the meantime.  

Within the deleter I cannot throw because it would happen during the destruction of the raii object. But I added an assertion.

Reply:
One could of course also argue that something is severely broken if you can create the directory but create a file inside (which is the only thing that would lead to abnormal termination and lacking clean-up). But of course it is good style to do it with RAII.

In general I wonder whether we shouldn't factor this out somewhere and/or whether you can re-use the temporary-file code we have in our tests.

Reply:
> In general I wonder whether we shouldn't factor this out somewhere and/or whether you can re-use the temporary-file code we have in our tests.

I looked into the temp file and it was not usable like it is there. But indeed it could have some advantages to make it a RAII class. Since the destructor cannot throw we can offer an explicit remove member function that throws. So normal code can remove the path but if anything else went south the destructor will silently cleanup.